weblab/repocache/populate.py

```diff
@@ -13,25 +11,26 @@ def populate_entity_cache(entity):
 
     :param entity: entity to process
     """
-    # The delete will cascade to versions and tags
-    CachedEntity.objects.filter(entity=entity).delete()
-
-    cached = CachedEntity.objects.create(entity=entity)
+    cached, cache_created = CachedEntity.objects.get_or_create(entity=entity)
 
     tag_dict = entity.repo.tag_dict
 
     visibility = entity.DEFAULT_VISIBILITY
     commits_without_visibility = []
+    valid_shas = set()
     for commit in entity.repo.commits:
+        valid_shas.add(commit.hexsha)
         visibility = entity.get_visibility_from_repo(commit)
-        version = CachedEntityVersion.objects.create(
+        version = CachedEntityVersion.objects.get_or_create(
             entity=cached,
             sha=commit.hexsha,
-            timestamp=commit.committed_at,
-            visibility=visibility or entity.DEFAULT_VISIBILITY
-        )
+            defaults={
+                'timestamp': commit.committed_at,
+                'visibility': visibility or entity.DEFAULT_VISIBILITY,
```

Comment:
This doesn't seem like it'll update the cache if the visibility of a version changes?

Reply:
Good spot, yes. I'll add a test case for that!